File: Projection/matrix.py
"""

3xN matrix object and operations are defined here.

"""


def dot(v3_1, v3_2):
    # On the assumption that len(v3_1) == len(v3_2),
    return sum([v3_1[i] * v3_2[i] for i in range(len(v3_1))])


class M3xN:
    def __init__(self, n):
        self.n = n
        self._elems = [[0 for _ in range(n)] for _ in range(3)]

    def set_entire(self, m):
        self._elems = m

    def set(self, row, col, val):
        self._elems[row][col] = val

    def get(self, row, col):
        return self._elems[row][col]

    def _get_row(self, i):
        return self._elems[i]

    def _get_col(self, j):
        col = []
        for r in self._elems:
            col.append(r[j])
        return col

    def __repr__(self):
        return str(self._elems)

    def __mul__(self, o):
        if type(o) != M3xN:
            res_mat = self
            for i in range(3):
                for j in range(self.n):
                    res_mat.set(i, j, self._elems[i][j] * o)

        else:
            res_mat = M3xN(o.n)
            for i in range(3):
                for j in range(o.n):
                    row = self._get_row(i)
                    col = o._get_col(j)
                    res_mat.set(i, j, dot(row, col))

        return res_mat

    def __add__(self, m):
        res_mat = self

        for i in range(3):
            for j in range(self.n):
                res_mat.set(i, j, self._elems[i][j] + m.get(i, j))

        return res_mat

    def __sub__(self, m):
        res_mat = self

        for i in range(3):
            for j in range(self.n):
                res_mat.set(i, j, self._elems[i][j] - m.get(i, j))

        return res_mat

File: Projection/matrix.cpp
#include "matrix.h"


float dot(vector<float> v1, vector<float> v2)
{
    float res = 0.0;
    for (int i = 0; i < v2.size(); i++)
    {
        res += v1[i] * v2[i];
    }
    return res;
}


MxN::MxN(int m, int n)
{
    m_m = m;
    m_n = n;
    
    for (int i = 0; i < m; i++)
    {
        vector<float> col;
        for (int j = 0; j < m; j++)
        {
            col.push_back(0.0);
        }

        m_elems.push_back(col);
    }

}


void MxN::print()
{
    for (auto r : m_elems)
    {
        for (auto e : r)
        {
            std::cout << e << ' ';
        }

        std::cout << '\n';
    }

    std::cout << '\n';
}


vector<float> MxN::m_get_row(int i)
{
    return m_elems[i];
}


vector<float> MxN::m_get_col(int j)
{
    vector<float> col;
    for (auto r : m_elems)
    {
        col.push_back(r[j]);
    }

    return col;
}



MxN MxN::operator *(MxN o)
{
    if (m_n == o.m_m)
    {
        MxN resMat(m_m, o.m_n);
        for (int i = 0; i < m_n; i++)
        {
            for (int j = 0; j < o.m_n; j++)
            {
                vector<float> row = m_get_row(i);
                vector<float> col = o.m_get_col(j);
                resMat.m_elems[i][j] = dot(row, col);
            }
        }

        return resMat;
    }

    else
    {
        std::cout << "Invalid multiplicaton" << std::endl;
    }
}


MxN MxN::operator +(MxN o)
{
    MxN resMat(m_m, m_n);
    for (int i = 0; i < m_n; i++)
    {
        for (int j = 0; j < m_m; j++)
        {
            resMat.m_elems[i][j] = m_elems[i][j] + o.m_elems[i][j];
        }
    }

    return resMat;
}


MxN MxN::operator -(MxN o)
{
    // TODO Hard coded 3x1 matrix sub
  
    MxN resMat(m_m, m_n);
    for (int i = 0; i < m_m; i++)
    {
        for (int j = 0; j < m_n; j++)
        {
            resMat.m_elems[i][j] = m_elems[i][j] - o.m_elems[i][j];
        }
    }

    return resMat;
}